orchagent/main.cpp

```diff
@@ -140,6 +145,121 @@ void init_gearbox_phys(DBConnector *applDb)
+{
+    DBConnector cfgDb("CONFIG_DB", 0);
+    DBConnector appDb("APPL_DB", 0);
+
+    Table cfgDeviceMetaDataTable(&cfgDb, CFG_DEVICE_METADATA_TABLE_NAME);
+    Table cfgSystemPortTable(&cfgDb, CFG_SYSTEM_PORT_TABLE_NAME);
+    Table appSystemPortTable(&appDb, APP_SYSTEM_PORT_TABLE_NAME);
+
+    //At this point of time (i.e, at the time of orchagent start), the CONFIG_DB is completely fully
+    //populated with the contents of config_db.json.
+
+    string value;
+    //Get the swith type.
+    if(!cfgDeviceMetaDataTable.hget("localhost", "switch_type", value))
+    {
+        //Switch type is not configured. Consider it default = "switch" (regular switch)
+        value = "switch";
+    }
+
+    if(value != "voq" && value != "fabric" && value != "switch")
+    {
+        cout << "Invalid switch type " << value.c_str() << " configured";
+        return false;
+    }
+    gMySwitchType = value;
+
+    if(gMySwitchType != "voq")
+    {
+        //Non VOQ switch. Nothing to read
+        return true;
+    }
+
+    if(!cfgDeviceMetaDataTable.hget("localhost", "switch_id", value))
+    {
+        //VOQ switch id is not configured.
+        cout << "VOQ switch id is not configured";
+        return false;
+    }
+
+    if(value.size())
+        gVoqMySwitchId = stoi(value);
+
+    if(gVoqMySwitchId < 0)
+    {
+        cout << "Invalid VOQ switch id " << gVoqMySwitchId << " configured";
+        return false;
+    }
+
+    if(!cfgDeviceMetaDataTable.hget("localhost", "max_cores", value))
+    {
+        //VOQ max cores is not configured.
+        cout << "VOQ max cores is not configured";
+        return false;
+    }
+
+    if(value.size())
+        gVoqMaxCores = stoi(value);
+
+    if(gVoqMaxCores == 0)
+    {
+        cout << "Invalid VOQ max cores " << gVoqMaxCores << " configured";
+        return false;
+    }
+
+    vector<string> spKeys;
+    cfgSystemPortTable.getKeys(spKeys);
+
+    //Retrieve system port configurations
+    vector<FieldValueTuple> spFv;
+    sai_system_port_config_t sysport;
+    for ( auto &k : spKeys )
+    {
+        cfgSystemPortTable.get(k, spFv);
+
+        for ( auto &fv : spFv )
+        {
+            if(fv.first == "switch_id")
+            {
+                sysport.attached_switch_id = stoi(fv.second);
+                continue;
+            }
+            if(fv.first == "core_index")
+            {
+                sysport.attached_core_index = stoi(fv.second);
+                continue;
+            }
+            if(fv.first == "core_port_index")
+            {
+                sysport.attached_core_port_index = stoi(fv.second);
+                continue;
+            }
+            if(fv.first == "speed")
+            {
+                sysport.speed = stoi(fv.second);
+                continue;
+            }
+            if(fv.first == "system_port_id")
+            {
+                sysport.port_id = stoi(fv.second);
+                continue;
+            }
```

Comment:
Why don't we check sysport num_voq setting in SystemPortTable?

Reply:
I can check for non-zero.
We can not check for a specific value since this value is optional and will be ignored by some vendor SAIs. I'll check the value for  0 and pass it to SAI as is.

Reply:
Fixed